compiler/rustc_mir/src/transform/deduplicate_blocks.rs

```diff
@@ -0,0 +1,161 @@
+        // with the error `funclet bb11 has 2 parents - bb24 and bb26`
+        .filter(|(_, bb_data)| !bb_data.is_cleanup)
+        .map(|(bb_idx, bb_data)| {
+            (
+                bb_idx,
+                Metadata {
+                    len_statements: bb_data.statements.len(),
+                    terminator_kind: bb_data.terminator().kind.clone(),
+                },
+            )
+        })
+        .collect();
+
+    // We do not use an unstable sort, as we want to preserve the order of the basic blocks.
+    // The current order is always increasing. This property is important,
+    // as we are using the same order when replacing basic blocks in terminators.
+    // With an ascending basic block index order, if bb1, bb2, and bb3 were all equal,
+    // we would replace bb1 with bb2, and then bb2 with bb3.
+    // If we used an unstable sort, we could potentially replace bb2 with bb3, and then bb1 with bb2.
+    // This would leave both bb3 and bb2, thereby not getting the full potential of this optimization.
+    to_sort.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
+    let iter = &to_sort.into_iter().group_by(|x| x.1.clone());
+    for x in iter {
+        let (group_key, group): (_, _) = x;
+        // We know by this point that the length of statements,
+        // and the TerminatorKind are equal within the group.
+        // Check that the statements are pairwise equal
+        let group: Vec<_> = group.collect();
+        debug!("group: `{:?}` ", group);
+        for (item1, item2) in group.iter().tuple_combinations() {
+            debug!("item1: `{:?}` item2: `{:?}`", item1, item2);
+            let statementkinds1 = body.basic_blocks()[item1.0].statements.iter().map(|x| &x.kind);
+            let statementkinds2 = body.basic_blocks()[item2.0].statements.iter().map(|x| &x.kind);
+
+            if statementkinds1.zip_eq(statementkinds2).all(|(lhs, rhs)| statement_eq(lhs, rhs)) {
+                debug!(
+                    "Inserting: key: `{:?}` value `{:?}`. group_key: `{:?}`",
+                    item1.0, item2.0, group_key,
+                );
+                duplicates.insert(item1.0, item2.0);
+            }
+        }
+    }
+
+    duplicates
+}
+
+fn statement_eq<'tcx>(lhs: &StatementKind<'tcx>, rhs: &StatementKind<'tcx>) -> bool {
+    let res = match (lhs, rhs) {
+        (
+            StatementKind::Assign(box (place, rvalue)),
+            StatementKind::Assign(box (place2, rvalue2)),
+        ) => place == place2 && rvalue_eq(rvalue, rvalue2),
+        (x, y) => x == y,
+    };
+    debug!("statement_eq lhs: `{:?}` rhs: `{:?}` result: {:?}", lhs, rhs, res);
+    res
+}
+
+fn rvalue_eq(lhs: &Rvalue<'tcx>, rhs: &Rvalue<'tcx>) -> bool {
+    let res = match (lhs, rhs) {
+        (
+            Rvalue::Use(Operand::Constant(box Constant { user_ty: _, literal, span: _ })),
```

Comment:
The only reason for not using == is to not compare span. See https://github.com/rust-lang/rust/pull/77549#discussion_r499462146
Should we have a compare_without_user_info? Not comparing span is an issue that most optimization passes would like to solve I guess

Reply:
This is a riddiculous footgun. I'll open an issue to discuss this

Reply:
https://github.com/rust-lang/rust/issues/77608